src/components/NewClusterModal.js

```diff
@@ -42,217 +43,428 @@ const imageValidationRegexp = /^[A-Za-z0-9]+[\w./-]+(?::\w[\w.-]+)?(?:@[\w+.-]+:
-    const { googleProject, runtimeName } = currentCluster
-
-    if (isStopRequired) {
-      notify('info', 'To be updated, your runtime will now stop, and then start. This will take 3-5 minutes.')
-    }
-
-    return onSuccess(
-      Ajax().Clusters.cluster(googleProject, runtimeName).update({
-        runtimeConfig: this.getRuntimeConfig()
+  /**
+   * Transform the new environment config into the shape of runtime config
+   * returned from leonardo. Specifically used as an input to cost calculation
+   * for potential new configurations.
+   */
+  getPendingRuntimeConfig() {
+    const { runtime: newRuntime } = this.getNewEnvironmentConfig()
+    return {
+      cloudService: newRuntime.cloudService,
+      ...(newRuntime.cloudService === cloudServices.GCE ? {
+        bootDiskSize: 50,
+        machineType: newRuntime.machineType,
+        ...(newRuntime.diskSize ? {
+          diskSize: newRuntime.diskSize
+        } : {})
+      } : {
+        masterMachineType: newRuntime.masterMachineType,
+        masterDiskSize: newRuntime.masterDiskSize,
+        numberOfWorkers: newRuntime.numberOfWorkers,
+        ...(newRuntime.numberOfWorkers && {
+          numberOfPreemptibleWorkers: newRuntime.numberOfPreemptibleWorkers,
+          workerMachineType: newRuntime.workerMachineType,
+          workerDiskSize: newRuntime.workerDiskSize
+        })
       })
-    )
-  }
-
-  hasStartUpScriptChanged() {
-    const { jupyterUserScriptUri, currentClusterDetails } = this.state
-    const originalJupyterUserScriptUri = currentClusterDetails?.jupyterUserScriptUri || ''
-    return jupyterUserScriptUri !== originalJupyterUserScriptUri
+    }
   }
 
-  hasImageChanged() {
-    const { selectedLeoImage, customEnvImage, currentClusterDetails } = this.state
-    const { imageUrl } = currentClusterDetails ? this.getImageUrl(currentClusterDetails) : ''
-    return !_.includes(imageUrl, [selectedLeoImage, customEnvImage])
+  /**
+  * Transform new environment config into the shape of a disk returned from Leo.
+  * Used as an input to cost calculation.
+  */
+  getPendingDisk() {
+    const { persistentDisk: newPersistentDisk } = this.getNewEnvironmentConfig()
+    return { size: newPersistentDisk.size, status: 'Ready' }
   }
 
-  //determines whether the changes are applicable for a call to the leo patch endpoint
-  //see this for a diagram of the conditional this implements https://drive.google.com/file/d/1mtFFecpQTkGYWSgPlaHksYaIudWHa0dY/view
-  //this function returns true for cases 2 & 3 in this diagram
-  canUpdate() {
-    const { currentCluster } = this.props
+  sendCloudEnvironmentMetrics() {
+    const { runtime: newRuntime, persistentDisk: newPersistentDisk } = this.getNewEnvironmentConfig()
+    const { runtime: oldRuntime, persistentDisk: oldPersistentDisk } = this.getOldEnvironmentConfig()
+    const newMachineType = newRuntime && (newRuntime.cloudService === cloudServices.GCE ? newRuntime.machineType : newRuntime.masterMachineType)
```

Comment:
yeah a helper might be good for this. The way `machineType` and `masterMachineType` are used interchangeably in the environment configs was making it tricky to inline the call to `findMachineType` (see just below this line) which is why I pulled it out like that, but I agree it could be streamlined!

Reply:
Upon consideration, I think leaving this as-is for now is the best option. We could pull out a function, but I don't think that the value of streamlining this block would outweigh the complication of having another function in an already long file - especially as it would only be useful for this particular instance, and it might be hard to name is so the purpose of the function is clear so I think it would impact the readability. So for those reasons I'm inclined to leave it alone for now, and potentially come back to refactor after this is merged